Code/Editor/EditorFramework/Assets/Implementation/AssetCurator.cpp

```diff
@@ -1553,39 +1496,112 @@ ezSubAsset* ezAssetCurator::GetSubAssetInternal(const ezUuid& assetGuid)
-      pFileStatus->m_Status = ezFileStatus::Status::Unknown;
+      sTemp = desc->m_sFileExtension;
+      sTemp.ToLower();
 
-      ezUuid guid0 = pFileStatus->m_AssetGuid;
-      if (guid0.IsValid())
+      AllExtensions.Insert(sTemp);
+    }
+  }
+}
+
+void ezAssetCurator::OnAssetFilesEvent(const ezFileChangedEvent& e)
+{
+  switch (e.m_Type)
+  {
+    case ezFileChangedEvent::Type::DocumentLinked:
+    {
+    }
+    break;
+    case ezFileChangedEvent::Type::FileAdded:
+    case ezFileChangedEvent::Type::FileChanged:
+    {
+      // If the asset was just added it is not tracked and thus no need to invalidate anything.
+      if (e.m_Type == ezFileChangedEvent::Type::FileChanged)
+      {
+        ezUuid guid0 = e.m_Status.m_DocumentID;
+        if (guid0.IsValid())
+          InvalidateAssetTransformState(guid0);
+
+        auto it = m_InverseTransformDeps.Find(e.m_sPath);
+        if (it.IsValid())
+        {
+          for (const ezUuid& guid : it.Value())
+          {
+            InvalidateAssetTransformState(guid);
+          }
+        }
+
+        auto it2 = m_InverseThumbnailDeps.Find(e.m_sPath);
+        if (it2.IsValid())
+        {
+          for (const ezUuid& guid : it2.Value())
+          {
+            InvalidateAssetTransformState(guid);
+          }
+        }
+      }
+
+      // Assets should never be in an AssetCache folder.
+      const char* szNeedle = e.m_sPath.FindSubString("AssetCache/");
+      if (szNeedle != nullptr && e.m_sPath.GetData() != szNeedle && szNeedle[-1] == '/')
       {
-        ezAssetInfo* pAssetInfo = m_KnownAssets[guid0];
-        UntrackDependencies(pAssetInfo);
-        RemoveAssetTransformState(guid0);
-        SetAssetExistanceState(*pAssetInfo, ezAssetExistanceState::FileRemoved);
-        pFileStatus->m_AssetGuid = ezUuid();
+        return;
       }
 
-      auto it = m_InverseTransformDeps.Find(sAbsolutePath);
+      // check that this is an asset type that we know
+      ezStringBuilder sExt = ezPathUtils::GetFileExtension(e.m_sPath);
+      sExt.ToLower();
+      if (!m_ValidAssetExtensions.Contains(sExt))
+      {
+        return;
+      }
+
+      // #TODO_ASSET -> wrong? This will update the timestamp for assets.
+      EnsureAssetInfoUpdated(e.m_sPath, e.m_Status).IgnoreResult();
+    }
+    break;
+    case ezFileChangedEvent::Type::FileRemoved:
+    {
+      ezUuid guid0 = e.m_Status.m_DocumentID;
+      if (guid0.IsValid())
+      {
+        // this is a bit tricky:
+        // when the document is deleted on disk, it would be nicer not to close it (discarding modifications!)
+        // instead we could set it as modified
+        // but then when it was only moved or renamed that means we have another document with the same GUID
+        // so once the user would save the now modified document, we would end up with two documents with the same GUID
+        // so, for now, since this is probably a rare case anyway, we just close the document without asking
```

Comment:
Did you test this recently? A long time ago I added this feature to close documents that were modified on disk, but recently I noticed that it often didn't work anymore. Does it work better now?

Reply:
I haven't tested it extensively no. EnsureDocumentIsClosedInAllManagers in itself should work fine. If it failed in the past it must have been due to bugs in the file tracking. This should be much more stable now (and tested).